src/ray/object_manager/test/pull_manager_test.cc

```diff
@@ -295,38 +322,193 @@ TEST_F(PullManagerTest, TestDeduplicateBundles) {
+  }
+  ASSERT_EQ(num_send_pull_request_calls_, oids.size());
+  ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+  AssertNumActiveRequestsEquals(oids.size());
+  ASSERT_TRUE(IsUnderCapacity(oids.size() * object_size));
+
+  // Reduce the available memory.
+  pull_manager_.UpdatePullsBasedOnAvailableMemory(oids.size() * object_size - 1);
+  AssertNumActiveRequestsEquals(0);
+  // No new pull requests after the next tick.
+  fake_time_ += 10;
+  auto prev_pull_requests = num_send_pull_request_calls_;
+  for (size_t i = 0; i < oids.size(); i++) {
+    pull_manager_.OnLocationChange(oids[i], client_ids, "", object_size);
+    ASSERT_EQ(num_send_pull_request_calls_, prev_pull_requests);
+    ASSERT_EQ(num_restore_spilled_object_calls_, 0);
+  }
+
+  // Increase the available memory again.
+  pull_manager_.UpdatePullsBasedOnAvailableMemory(oids.size() * object_size);
+  AssertNumActiveRequestsEquals(oids.size());
+  ASSERT_TRUE(IsUnderCapacity(oids.size() * object_size));
+  ASSERT_EQ(num_send_pull_request_calls_, prev_pull_requests + oids.size());
+
+  pull_manager_.CancelPull(req_id);
+  AssertNumActiveRequestsEquals(0);
+}
+
+TEST_F(PullManagerWithAdmissionControlTest, TestQueue) {
+  /// Test admission control for a queue of pull bundle requests. We should
+  /// activate as many requests as we can, subject to the reported capacity.
+  int object_size = 2;
+  int num_oids_per_request = 2;
+  int num_requests = 3;
+
+  std::vector<std::vector<ObjectID>> bundles;
+  std::vector<int64_t> req_ids;
+  for (int i = 0; i < num_requests; i++) {
+    auto refs = CreateObjectRefs(num_oids_per_request);
+    auto oids = ObjectRefsToIds(refs);
+    std::vector<rpc::ObjectReference> objects_to_locate;
+    auto req_id = pull_manager_.Pull(refs, &objects_to_locate);
+    ASSERT_EQ(ObjectRefsToIds(objects_to_locate), oids);
+
+    bundles.push_back(oids);
+    req_ids.push_back(req_id);
+  }
+
+  std::unordered_set<NodeID> client_ids;
+  client_ids.insert(NodeID::FromRandom());
+  for (auto &oids : bundles) {
+    for (size_t i = 0; i < oids.size(); i++) {
+      pull_manager_.OnLocationChange(oids[i], client_ids, "", object_size);
+    }
+  }
+
+  for (int capacity = 0; capacity < 20; capacity++) {
+    int num_requests_expected =
+        std::min(num_requests, capacity / (object_size * num_oids_per_request));
+    pull_manager_.UpdatePullsBasedOnAvailableMemory(capacity);
+
+    AssertNumActiveRequestsEquals(num_requests_expected * num_oids_per_request);
+    // The total requests that are active is under the specified capacity.
+    ASSERT_TRUE(
+        IsUnderCapacity(num_requests_expected * num_oids_per_request * object_size));
+    // This is the maximum number of requests that can be served at once that
+    // is under the capacity.
+    if (num_requests_expected < num_requests) {
+      ASSERT_FALSE(IsUnderCapacity((num_requests_expected + 1) * num_oids_per_request *
+                                   object_size));
+    }
+  }
```

Comment:
Can you also test the case where the whole pull requests are digested and new pull requests are queued to test this part? 
``` cpp
if (highest_req_id_being_pulled_ == request_it->first) {
    if (request_it == pull_request_bundles_.begin()) {
      highest_req_id_being_pulled_ = 0;
    } else {
      highest_req_id_being_pulled_ = std::prev(request_it)->first;
```

Reply:
Hmm I think this will get covered by the cancellation tests, but let me know if there is something else you're looking for.

Reply:
Ah, if it is convered by that test, it is fine. (It was a bit hard to understand if this was tested).